pkg/sac/allow_fixed_scopes_checker_core.go

```diff
@@ -3,121 +3,256 @@ package sac
+		clusterKeys:   typedKeySet[ClusterScopeKey]([]ScopeKey{}),
+		namespaceKeys: typedKeySet[NamespaceScopeKey]([]ScopeKey{}),
 	}
-	return denyAllScopeCheckerCore
 }
 
-func (c allowFixedScopesCheckerCore) EffectiveAccessScope(resource permissions.ResourceWithAccess) (*effectiveaccessscope.ScopeTree, error) {
-	if len(c) == 0 {
-		return effectiveaccessscope.UnrestrictedEffectiveAccessScope(), nil
+func allowFixedClusterLevelScopes(
+	accessLevelKeys []ScopeKey,
+	resourceLevelKeys []ScopeKey,
+	clusterLevelKeys []ScopeKey,
+) ScopeCheckerCore {
+	return &allowedFixedScopesCheckerCore{
+		checkerLevel:  GlobalScopeKind,
+		accessKeys:    typedKeySet[AccessModeScopeKey](accessLevelKeys),
+		resourceKeys:  typedKeySet[ResourceScopeKey](resourceLevelKeys),
+		clusterKeys:   typedKeySet[ClusterScopeKey](clusterLevelKeys),
+		namespaceKeys: typedKeySet[NamespaceScopeKey]([]ScopeKey{}),
 	}
-	// TODO (ROX-9981): Change the structure of allowFixedScopesCheckerCore to have explicit scope level semantic
-	for key := range c.topLevelKeys() {
-		switch key.(type) {
-		case AccessModeScopeKey:
-			return c.getAccessModeEffectiveAccessScope(resource)
-		case ResourceScopeKey:
-			return c.getResourceEffectiveAccessScope(resource)
-		case ClusterScopeKey:
-			return c.getClusterEffectiveAccessScope()
+}
+
+func allowFixedNamespaceLevelScopes(
+	accessLevelKeys []ScopeKey,
+	resourceLevelKeys []ScopeKey,
+	clusterLevelKeys []ScopeKey,
+	namespaceLevelKeys []ScopeKey,
+) ScopeCheckerCore {
+	return &allowedFixedScopesCheckerCore{
+		checkerLevel:  GlobalScopeKind,
+		accessKeys:    typedKeySet[AccessModeScopeKey](accessLevelKeys),
+		resourceKeys:  typedKeySet[ResourceScopeKey](resourceLevelKeys),
+		clusterKeys:   typedKeySet[ClusterScopeKey](clusterLevelKeys),
+		namespaceKeys: typedKeySet[NamespaceScopeKey](namespaceLevelKeys),
+	}
+}
+
+func (c *allowedFixedScopesCheckerCore) SubScopeChecker(scopeKey ScopeKey) ScopeCheckerCore {
+	switch key := scopeKey.(type) {
+	case AccessModeScopeKey:
+		if c.checkerLevel != GlobalScopeKind {
+			return denyAllScopeCheckerCore
+		}
+		if c.accessKeys.Cardinality() > 0 && !c.accessKeys.Contains(key) {
+			return denyAllScopeCheckerCore
+		}
+		return &allowedFixedScopesCheckerCore{
+			checkerLevel:  AccessModeScopeKind,
+			accessKeys:    c.accessKeys,
+			resourceKeys:  c.resourceKeys,
+			clusterKeys:   c.clusterKeys,
+			namespaceKeys: c.namespaceKeys,
+			accessLevel:   key,
+		}
+	case ResourceScopeKey:
+		if c.checkerLevel != AccessModeScopeKind {
+			return denyAllScopeCheckerCore
+		}
+		if c.resourceKeys.Cardinality() > 0 && !c.resourceKeys.Contains(key) {
+			return denyAllScopeCheckerCore
+		}
+		return &allowedFixedScopesCheckerCore{
+			checkerLevel:   ResourceScopeKind,
+			accessKeys:     c.accessKeys,
+			resourceKeys:   c.resourceKeys,
+			clusterKeys:    c.clusterKeys,
+			namespaceKeys:  c.namespaceKeys,
+			accessLevel:    c.accessLevel,
+			targetResource: key,
+		}
```

Comment:
If I understna correctly on given level unsued fields are empty. How about using builder pattern (or slightly modified version of it). If we have following functions
```go
func (c *allowedFixedScopesCheckerCore) subScopeChecker() *allowedFixedScopesCheckerCore {
	return &allowedFixedScopesCheckerCore{
		checkerLevel:  c.checkerLevel + 1,
		accessKeys:      c.accessKeys,
		resourceKeys:    c.resourceKeys,
		clusterKeys:     c.clusterKeys,
		namespaceKeys:   c.namespaceKeys,
		accessLevel:     c.accessLevel,
		targetResource:  c.targetResource,
		targetCluster:   c.targetCluster,
	}
}


func (c *allowedFixedScopesCheckerCore) withAccessLevel(key AccessModeScopeKey) *allowedFixedScopesCheckerCore {
	c.accessLevel = key
	return c
}
func (c *allowedFixedScopesCheckerCore) withTargetResource(key ResourceScopeKey) *allowedFixedScopesCheckerCore {
	c.targetResource = key
	return c
}
func (c *allowedFixedScopesCheckerCore) withTargetCluster(key ClusterScopeKey) *allowedFixedScopesCheckerCore {
	c.targetCluster = key
	return c
}
func (c *allowedFixedScopesCheckerCore) withTargetNamespace(key NamespaceScopeKey) *allowedFixedScopesCheckerCore {
	c.targetNamespace = key
	return c
}
```

then this line could be sipilified to 
```suggestion
		return c.subScopeChecker().withTargetResource(key)
```